admin/server/postgres.go

```diff
@@ -0,0 +1,204 @@
+		password := ctx.Value(auth.PostgresPassword{}).(string)
+		if len(password) >= 6 && strings.EqualFold(password[0:6], "bearer") {
+			jwt = strings.TrimSpace(password[6:])
+		}
+	case auth.OwnerTypeUser, auth.OwnerTypeService:
+		// If the client is authenticated with the admin service, we issue a new ephemeral runtime JWT.
+		// The JWT should have the same permissions/configuration as one they would get by calling AdminService.GetProject.
+		permissions := claims.ProjectPermissions(ctx, proj.OrganizationID, depl.ProjectID)
+		if !permissions.ReadProd {
+			return nil, fmt.Errorf("does not have permission to access the production deployment")
+		}
+
+		var attr map[string]any
+		if claims.OwnerType() == auth.OwnerTypeUser {
+			attr, err = s.jwtAttributesForUser(ctx, claims.OwnerID(), proj.OrganizationID, permissions)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		jwt, err = s.issuer.NewToken(runtimeauth.TokenOptions{
+			AudienceURL: depl.RuntimeAudience,
+			Subject:     claims.OwnerID(),
+			TTL:         runtimeAccessTokenDefaultTTL,
+			InstancePermissions: map[string][]runtimeauth.Permission{
+				depl.RuntimeInstanceID: {
+					// TODO: Remove ReadProfiling and ReadRepo (may require frontend changes)
+					runtimeauth.ReadObjects,
+					runtimeauth.ReadMetrics,
+					runtimeauth.ReadProfiling,
+					runtimeauth.ReadRepo,
+					runtimeauth.ReadAPI,
+				},
+			},
+			Attributes: attr,
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("runtime proxy not available for owner type %q", claims.OwnerType())
+	}
+
+	// Track usage of the deployment
+	s.admin.Used.Deployment(depl.ID)
+
+	hostURL, err := url.Parse(depl.RuntimeHost)
+	if err != nil {
+		return nil, err
+	}
+	hostURL.Scheme = "postgres"
+	hostURL.Host = hostURL.Hostname() + ":" + strconv.FormatInt(int64(15432), 10)
+	hostURL.User = url.UserPassword("postgres", fmt.Sprintf("Bearer %s", jwt))
+	hostURL.Path = depl.RuntimeInstanceID
+	conn, err := connectionPool(ctx, hostURL.String())
+	if err != nil {
+		s.logger.Info("error in get connection pool", zap.Error(err))
+		return nil, err
+	}
+
+	rows, err := conn.Query(ctx, query) // query to underlying host
+	if err != nil {
+		s.logger.Info("error in query", zap.Error(err))
+		return nil, err
+	}
+	defer rows.Close()
+
+	// handle schema
+	var cols []wire.Column
+	fds := rows.FieldDescriptions()
+	for _, fd := range fds {
+		cols = append(cols, wire.Column{
+			Table: int32(fd.TableOID),
+			Name:  fd.Name,
+			Oid:   oid.Oid(fd.DataTypeOID),
+			Width: fd.DataTypeSize,
+			Attr:  int16(fd.TableAttributeNumber),
+		})
+	}
+
+	// handle data
+	// NOTE :: This creates a copy of data and stores this till client starts reading data. This is required so that we
+	// can close runtime connection and not wait for client to complete reading whole data which can leak connection.
+	// We can improve this logic in future.
```

Comment:
Moving `rows.Close()` will mean connection is not freed till client reads whole data which we are trying to avoid here. This will also mean keeping 10 connections in memory for a client that initiates 10 queries but does not close connection properly. Given our runtime connections are multiplied by a factor of `org * project` we may be looking at a big number here.

Its essentially a tradeoff IMO:
1. Risk opening and keeping lot of open connections to runtime.
2. Buffer all data in-memory.

If you feel its better to keep the connection and do not buffer data, I can also make that change.

Reply:
> Given our runtime connections are multiplied by a factor of org * project we may be looking at a big number here. (not correct so may be not a big number)

Yes connections are unique for each instanceID so a big number. :)

Reply:
It's worth optimizing on the number of connections that are open, but a connection is likely cheaper than memory allocation and garbage collection for each cell of the output data. The connection is also likely to be kept open anyway (due to pools), and proxying directly instead of buffering is in line with the behavior usually expected of a proxy.